### PR TITLE
add a node cache to the UUIDBaseNode. This node cache is used to upda…

### PR DESCRIPTION
**Pull request by InnocentBug**
…te existing UUID nodes, rather then creating a new node with the same UUID.

# Description

## Checklist

- [ ] My name is on the list of contributors (`CONTRIBUTORS.md`) in the pull request source branch.
- [ ] I have updated the documentation to reflect my changes.

### CODE REVIEW

File: src/cript/nodes/core.py [hunk 1]
```diff
@@ -166,13 +166,20 @@ def _from_json(cls, json_dict: dict):
...
+
+        if "uuid" in json_dict and json_dict["uuid"] in UUIDBaseNode._uuid_cache:
+            node = UUIDBaseNode._uuid_cache[json_dict["uuid"]]
+        else:  # Create a new node
+            try:
+                node = cls(**arguments)
+            # TODO we should not catch all exceptions if we are handling them, and instead let it fail
+            #  to create a good error message that points to the correct place that it failed to make debugging easier
+            except Exception as exc:
+                print(cls, arguments)
```
Comment (nh916) [inline src/cript/nodes/core.py:181]: does this need to be removed?

File: src/cript/nodes/core.py [hunk 2]
```diff
@@ -166,13 +166,20 @@ def _from_json(cls, json_dict: dict):
...
-            print(cls, arguments)
-            raise exc
+        # If a node with this UUID already exists, we don't create a new node.
+        # Instead we use the existing node from the cache and just update it.
+        from cript.nodes.uuid_base import UUIDBaseNode
+
+        if "uuid" in json_dict and json_dict["uuid"] in UUIDBaseNode._uuid_cache:
+            node = UUIDBaseNode._uuid_cache[json_dict["uuid"]]
+        else:  # Create a new node
+            try:
```
Comment (nh916) [inline src/cript/nodes/core.py:176]: I see you check for cache hit, but in the case of cache miss are you adding it to the cache?

Comment (InnocentBug) [reply src/cript/nodes/core.py:176]: that happens automatically in the UUIDBaseNode constructor

File: src/cript/nodes/core.py (hunk 1 above)
Comment (InnocentBug) [reply src/cript/nodes/core.py:181]: maybe, we had it in there even before I changed it

**Review by nh916 — APPROVED** [review]

**Pull request merged**